### DISCUSSION TITLE
How to extract U-values of walls roof and windows from an IFC file using Python

### DISCUSSION THREAD

**Discussion by talha-siddiqui-7** [IfcOpenShell/IfcOpenShell#2746] [Q&A]
Hello everyone,

I am currently working on a project in which i have to use BIM data to develop a recommender system. At the moment i am trying to extract useful building information from the IFC file like the U-values of the envelope elements (walls, roofs, etc). However, i am unable to correctly do it. I have also tried some code from ChatGPT but receiving errors like:

Warning: u_value not found for wall with ID <bound method entity_instance.id of #10002=IfcWall('3XaUjkL_z2kA00GrJt5U$_',#20,'Basic Wall:My exterior wall 280mm:1026504:My exterior wall 280mm:1026504 : Basic Wall:My exterior wall 280mm:1026504:324654',$,'Basic Wall:My exterior wall 280mm:1026504:My exterior wall 280mm:1026504 : Basic Wall:My exterior wall 280mm:1026504',#9681,#10001,'324654')>
The wall's related property definition is: #10022=IfcPropertySet('1h76JlkpUI835OhKkeCPbh',#20,'Pset_QuantityTakeOff',$,(#10013))

the code i used was :

`import ifcopenshell

# Load the IFC file
ifc_file = ifcopenshell.open("D:/EMJMD-SMACCs/Thesis/Thesis work/Data/SmartHome_With U values.ifc")

# Get all the wall elements in the IFC file
wall_elements = [element for element in ifc_file.by_type("IfcWall")]

# Extract the U-values of the walls
u_values = []
for wall in wall_elements:
    # Check if the U-value is present in the RelatingPropertyDefinition
    if hasattr(wall.IsDefinedBy[0].RelatingPropertyDefinition, "UValue"):
        u_value = wall.IsDefinedBy[0].RelatingPropertyDefinition.UValue
        u_values.append(u_value)
    else:
        # Check if the U-value is present in the properties of the IfcPropertySet
        property_set = wall.IsDefinedBy[0].RelatingPropertyDefinition
        for prop in property_set.HasProperties:
            if prop.Name == "UValue":
                u_values.append(prop.NominalValue.value)
                break
        else:
            print("Warning: u_value not found for wall with ID", wall.id)

print(u_values)`

Can someone please help me with this. At the moment i am trying to study the the structure of IFC file. But i know it contains U-values as i myself put them there

**Comment by Moult** [comment]
Have you got a model you can share?

**Reply by talha-siddiqui-7** [reply]
i do not have a model yet. At the  moment i am just trying to see if i can extract useful information from the IFC file. For recommender system it self i will use a collaborative approach. But i need to be able to extract the useful information from the BIM file

**Reply by talha-siddiqui-7** [reply]
If you mean the IFC BIM file, that i can share here
[IFC BIM File.zip](https://github.com/IfcOpenShell/IfcOpenShell/files/10679505/IFC.BIM.File.zip)

**Reply by Moult** [reply]
U Values are stored as the "thermal transmittance" property for walls in the Pset_WallCommon property group. http://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/Pset_WallCommon.htm

Here is sample code of how to get properties from an element: https://blenderbim.org/docs-python/ifcopenshell-python/code_examples.html#get-the-properties-of-a-wall-type

**Reply by talha-siddiqui-7** [reply]
Thank you very much. It worked:)
